Clean up genericLongAndAsRotateHelper

dchopra001 opened the pull request:
-Rename the function to tryToReplaceLongAndWithRotateInstruction
as this is more descriptive

-Clean up the function to avoid skipping the evaluation of i2l
and iu2l nodes. This should be handled by the simplifier (and
it currently is). Also, insert a static debug counter to catch
potential opportunities in the future.

-Add asserts during binary encoding and instruction construction
to make sure the reserved bits in the instruction's operands are
not being set. This is a source of printing errors in trace logs.

-Test the function using tril + google tests

Signed-off-by: Dhruv Chopra <Dhruv.C.Chopra@ibm.com>

mgaudet commented on fvtest/compilertriltest/LongAndAsRotateTest.cpp:
```diff
@@ -0,0 +1,494 @@
+
+   auto entry_point = compiler.getEntryPoint<uint64_t (*)(int32_t)>();
+
+   ASSERT_EQ(param.oracle((param.variableValue & 0x4321), param.constantValue), entry_point(param.variableValue)); 
```
Where's the 0x4321 coming from?

dchopra001 replied:
I should have defined this more clearly. The 0x4321 is just a positive value that I use for the `iconst` node to get the optimizer to set the `i2l` node's highWordZero flag. 
The comment `//basic block extension will realize that this i2l will have highWordZero` was entered in relation to this. I'll make this more clear.

mgaudet replied:
Would it be useful do you think to be able to directly set the 'highWordZero' flag on a node? If so, maybe open an enhancement issue. It's something milling about in my mental backlog, but I don't have an explicit issue for it yet

dchopra001 replied:
I opened an issue: https://github.com/eclipse/omr/issues/1947

mgaudet commented on fvtest/compilertriltest/LongAndAsRotateTest.cpp:
```diff
@@ -0,0 +1,494 @@
+
+   static void SetUpTestCase()
+      {
+      const char *options = "-Xjit:disableTreeSimplification,disableAsyncCompilation,acceptHugeMethods,enableBasicBlockHoisting,omitFramePointer,useIlValidator,paranoidoptcheck";
```
I wonder if this test case wouldn't be a good candidate to test out the MockOptimization strategies, just to give you some more direct control over the strategy. (See #1661 and [this example](https://github.com/mgaudet/omr/blob/3049c28624e9b989d289687e28a615cba201c502/fvtest/compilertriltest/SimplifierFoldAndTest.cpp#L69-L84))

dchopra001 replied:
Interesting. I remember wishing there was something like this when I was testing edge cases. Some test cases were particularly designed at the time to test under different optimization settings. Not sure if it's needed now though, as we made some changes such that the testing was no longer needed. But I'll look into this.

dchopra001 replied:
So I spoke to Filip about this and realized I need to rethink the optimization strategy for these tests. In most cases we should not be having any optimizations in the strategy because these trees will generally be folded by the optimizer. I will enable the trees in certain cases when I want the optimizer to detect certain patterns and set appropriate flags. Perhaps we can also revisit this when we have a feature in tril that allows us to set flags (as you suggested in another comment here). I'll open an issue relating to this today.

dchopra001 replied:
So I played around with MockOptimization strategies, and I can make something work using that as well. However I still don't think this is the best long term solution. (Also, the segment of code I was trying to test will be removed in the next commit, but we'll ignore that for now as it can still be important to test this corner case).

The JIT code modified here is in the code generator and ideally I would like to run with noOpt to test the code. However I can't do that right now because some of the nodes require certain flags to be set. So my work around was to disable just enough optimizations so that the optimizer will set the flag I want. The way to do this with MockOptimization would be to run with noOpts except for a couple of test cases (where we would have to add then remove an optimization, or run these tests under a different TestFixture.

The best long term solution IMO here is to not use any optimization and set the flags myself. However the feature to set flags is not ready yet (https://github.com/eclipse/omr/issues/1947). I suggest we leave this as is and revisit these tests once we have the ability to set flags. Neither of the current solutions are particularly "cleaner" and are not the long term solution. So it doesn't really matter which one we go with right now. We will have to revisit this regardless. 

Let me know what you think.

mgaudet commented on fvtest/compilertriltest/LongAndAsRotateTest.cpp:
```diff
@@ -0,0 +1,494 @@
+   static void SetUpTestCase()
+      {
+      const char *options = "-Xjit:disableTreeSimplification,disableAsyncCompilation,acceptHugeMethods,enableBasicBlockHoisting,omitFramePointer,useIlValidator,paranoidoptcheck";
+      auto initSuccess = initializeJitWithOptions(const_cast<char*>(options));
```
Can you open an issue to change the signature of `initializeJitWithOptions` so that this `const_cast` becomes un-necessary?

dchopra001 replied:
I'm guessing you already got to it here?
https://github.com/eclipse/omr/issues/1916

mgaudet replied:
Actually, that's a separate issue entirely

dchopra001 replied:
Done: https://github.com/eclipse/omr/issues/1946

mgaudet commented:
(As an aside, though I wouldn't have you change on this basis alone: It seems like your commit message is wrapped at ~40 chars... which is weirdly narrow. CONTRIBUTING.md says 72 is a best effort IIRC)

0xdaryl commented:
I'd appreciate a review from @fjeremic please.

0xdaryl commented:
I haven't reviewed yet, but @dchopra001 could you fix the formatting in the commit and PR message as @mgaudet suggests and rebase to fix the merge conflict.

dchopra001 commented on fvtest/compilertriltest/LongAndAsRotateTest.cpp:
```diff
@@ -0,0 +1,494 @@
+         "   )                                                 "
+         ")                                                    ",  
+         std::to_string(param.constantValue).c_str()
+ */       //above tree is compacted below:
```
I thought the extra white spaces was wasteful :). Let me know if it's better to stick to the original form though.

mgaudet replied:
But now you have both the compacted and original forms 😄 

I heavily prefer the whitespace laden form as it's more readable. I'd only keep it around unless there's a compelling reason the compacted form is better.

fjeremic commented on compiler/z/codegen/BinaryEvaluator.cpp:
```diff
@@ -2466,8 +2466,36 @@ genericRotateAndInsertHelper(TR::Node * node, TR::CodeGenerator * cg)
+ * to preserve the first 36 bits (lconst's first 36 bits are 1's) and 
+ * accomplish this in one instruction.  
+ *
+ */
```
Can we properly Doxygen this function an its parameters in the header and leave the details Doxygen'd in the object file implementation?

fjeremic commented on compiler/z/codegen/BinaryEvaluator.cpp:
```diff
@@ -2466,8 +2466,36 @@ genericRotateAndInsertHelper(TR::Node * node, TR::CodeGenerator * cg)
+ *    lconst 0xFFFFFFFFF0000000
+ *
+ * We can avoid the 'land' operation altogether in such a case. We know that
+ * the result of this operation will preserve the first 16 bits of the iload
```
> will preserve the first 16 bits of the iload

This does not sound correct. This example is kind of an edge case. I would focus more on the intention of the use of `RISBG` to avoid two AND instructions, i.e. the case where you are ANDing with something like `0xFF00000000000003` where there is a contiguous sequence of one bits somewhere in the constant and where the alternative to evaluating such a sequence is to either materialize the constant in a register (expensive) or to use two ANDs such as `NIHF` + `NILF` to and the most-significant and least-significant 32-bit portions of the first operand.

fjeremic commented on compiler/z/codegen/BinaryEvaluator.cpp:
```diff
@@ -2476,56 +2504,75 @@ genericLongAndAsRotateHelper(TR::Node * node, TR::CodeGenerator * cg)
       TR::Node * skipedConversion = NULL;
```
There doesn't seem to be any uses of this variable. Are we able to fold it away?

fjeremic commented on compiler/z/codegen/BinaryEvaluator.cpp:
```diff
@@ -2476,56 +2504,75 @@ genericLongAndAsRotateHelper(TR::Node * node, TR::CodeGenerator * cg)
+         uint64_t longConstValue = secondChild->getLongInt();
 
          if ((firstChild->getOpCodeValue() == TR::i2l || firstChild->getOpCodeValue() == TR::iu2l)
                && firstChild->getReferenceCount()==1 && firstChild->getRegister() == NULL && firstChild->isHighWordZero() )
```
`firstChild->getReferenceCount()==1 && firstChild->getRegister() == NULL` can be replaced by `firstChild->isSingleRefUnevaluated()`. Why does the reference count matter here?

dchopra001 replied:
So the way I understand is that we would originally skip the evaluation of the `i2l` or `iu2l` node if this entire condition holds true. `firstChild` would be set to firstChild->getFirstChild() and if a RISBG instruction can be generated, then we wouldn't evaluate the `i2l` or `iu2l` node anymore. If the reference count is greater than 1 then we cannot skip the evaluation because we might need the result later on. 

I left this check here because my initial thought was that a case like this should never reach the code generator because the optimizer should have further simplified a tree like this. Is this not true? (I can see why the name of the counter can be misleading. I'll try coming up with a better one).

mgaudet replied:
(_Outsider Opinion_ If it's expected that the optimizer **must** have done a simplification, we need to documented in some fashion to avoid tripping people up when doing investigations, or when disabling some or all of the optimizer. Right now, there's not a great home for this, but we're working on developing a 'pre-codegen validation' which is where you could sock these kinds of facts away. See #1482 )

fjeremic replied:
> So the way I understand is that we would originally skip the evaluation of the i2l or iu2l node if this entire condition holds true.

The function `tryToReplaceLongAndWithRotateInstruction` as it stands does not skip the evaluation of any children. If a truncation operation should not happen a simplifier case should be added to eliminate it. The function implemented here is responsible for determining if an evaluation of an AND node can benefit from being evaluated with an `RISBG` instruction. The reference count of the children of the AND operation should not matter here. Is the logic here correct?

fjeremic commented on compiler/z/codegen/BinaryEvaluator.cpp:
```diff
@@ -2476,56 +2504,75 @@ genericLongAndAsRotateHelper(TR::Node * node, TR::CodeGenerator * cg)
-            firstChild = firstChild->getFirstChild();
-            value &= 0x00000000FFFFFFFFL;
-            }
+            TR::DebugCounter::incStaticDebugCounter(cg->comp(), TR::DebugCounter::debugCounterName(cg->comp(), "codegen/z/BinaryEvaluator/simplifierShouldHaveFoldedThisTree/%s", cg->comp()->signature()));
```
This is not true. The tree should have been folded by the simplifier if and only if `value & 0x00000000FFFFFFFFL == 0`. We should not blindly increment the counter here without such a check.

dchopra001 replied:
So based on our discussion earlier my understanding was that this type of code should be in the optimizer. The optimizer would/should be able to simplify a tree that looks like this, and not just for the specific case where `value & 0x00000000FFFFFFFFL == 0`. So we would leave a check like this here because it could help identify future opportunities. Is this not true?

fjeremic replied:
That is only partially correct. Consider the following tree:

```
land
  i2l (highWordZero)
    ...
  lconst 1
```

This tree would enter the above `if` statement and your debug counter would be incremented. However this tree cannot be folded by the simplifier because `value & 0x00000000FFFFFFFFL != 0`, i.e. you have to carry out the AND operation. So the debug counter spots a false positive.

I'm more inclined to say we just remove this entire `if` all together. I don't particularly see much value in it.

fjeremic commented on compiler/z/codegen/BinaryEvaluator.cpp:
```diff
@@ -2476,56 +2504,75 @@ genericLongAndAsRotateHelper(TR::Node * node, TR::CodeGenerator * cg)
             lsBit = 63 - tZeros;
-            if (firstChild->getReferenceCount()>1 || skipedConversion != NULL   // Better of with non-destructive RISBG as AND?
-                  || lZeros > 31 || tZeros > 31 || (lZeros > 0 && tZeros > 0))  // Better off with a single AND?
+            if (firstChild->getReferenceCount() > 1 // if we use an AND and refCount>1, then we will have to backup the value in another register beforehand, as AND is destructive
```
Can't we clobber evaluate `firstChild` and perform the transformation anyway? Can you further explain why reference count matters here?

dchopra001 replied:
So the RISBG instruction is non-destructive and we can keep the value in a register as is. So if the reference count > 1 we can generate something like this: `RISBGN R1, R2, 0,129,0` in order to preserve the data in R2. I don't understand clobberEvaluate completely, but from what I understand so far it would keep a copy of the data in R2 in a separate register. Isn't this unnecessary since we can use `RISBGN` to not clobber the value in the first place?

This is what my comment on the line kind of refers to as well. If I'm missing something here then let me know. Otherwise I think I'll add a better comment in the code to document this.

fjeremic replied:
> So the RISBG instruction is non-destructive and we can keep the value in a register as is. So if the reference count > 1 we can generate something like this: RISBGN R1, R2, 0,129,0 in order to preserve the data in R2. I don't understand clobberEvaluate completely, but from what I understand so far it would keep a copy of the data in R2 in a separate register. Isn't this unnecessary since we can use RISBGN to not clobber the value in the first place?

That's correct. Then there is no need to clobber evaluate in this case at all. I don't see why the reference count would then be a pre-requisite for this optimization?

dchopra001 replied:
If we were to remove the `firstChild->getReferenceCount()>1` condition, the inner if clause will become: `if (lZeros > 31 || tZeros > 31 || (lZeros > 0 && tZeros > 0)`. 

We can still have a case where the outer if clause (  `if (popCnt == (64 - lZeros - tZeros))`) is true, but the new inner one is not (ex. if `lConstValue` is originally `0x0000FFFFFFFFFFFF`). In this scenario we would bail from the helper and end up generating an `NIHH` instruction. If the firstChild's refCount>1, we will need to clobber evaluate this case because we cannot destroy the value. Wouldn't this result in an extra instruction since we need to backup the original value somewhere? 
If we went with a `RISBG` instruction instead, we can accomplish this in one instruction since we don't need to clobber evaluate. The number of registers used would be the same in both cases. This is why I believe if the refCount>1 for the firstChild we should always use RISBG since we save an instruction.

I'm going to test this case out once I get past the compile failures I'm currently seeing. Let me know if I'm missing something here.

fjeremic replied:
> We can still have a case where the outer if clause ( if (popCnt == (64 - lZeros - tZeros))) is true, but the new inner one is not (ex. if lConstValue is originally 0x0000FFFFFFFFFFFF). In this scenario we would bail from the helper and end up generating an NIHH instruction. If the firstChild's refCount>1, we will need to clobber evaluate this case because we cannot destroy the value. Wouldn't this result in an extra instruction since we need to backup the original value somewhere?

Right, but we miss all the opportunities where `lConstValue` is not one where we can `NIHH` for example. The point I'm trying to drive is that `firstChild->getReferenceCount() == 1` does not always imply that using `RISBG` is a bad idea. 

What if `firstChild->getReferenceCount() == 1` and `lConstValue == 0x0000FFFFFFFFFFF0` for example? This function would choose not to generate an `RISBG` and the AND evaluator will have to generate at least 2 ANDs.

fjeremic replied:
> We can still have a case where the outer if clause ( if (popCnt == (64 - lZeros - tZeros))) is true, but the new inner one is not (ex. if lConstValue is originally 0x0000FFFFFFFFFFFF). In this scenario we would bail from the helper and end up generating an NIHH instruction.

The difference between generating `NIHH` and `RISBG` in this case is that `RISBG` is 2 bytes longer. The alternative to not generating `RISBG` in case of `firstChild->getReferenceCount() == 1` and `lConstValue == 0x0000FFFFFFFFFFF0` is that multiple instructions will get generated, which is always worse than the extra 2 bytes of icache we pay.

The best of both worlds is to determine in which case can we *NOT* use one of the `NI**` variants to evaluate the sequence when `firstChild->getReferenceCount() == 1` and generate an `RISBG` in such cases (if possible of course).

dchopra001 replied:
Okay so I think we're both trying to say the same thing here.   

>The point I'm trying to drive is that firstChild->getReferenceCount() == 1 does not always imply that using RISBG is a bad idea.

 The `firstChild->getReferenceCount() > 1` check is not a prerequisite. The entire if clause looks like the following in a single line: 
`if (firstChild->getReferenceCount() > 1 || lZeros > 31 || tZeros > 31 || (lZeros > 0 && tZeros > 0))`

I separated these into multiple lines to add comments explaining what each one is doing. I'll clean this up and put a comment above the `if` to make this more clear.

>What if firstChild->getReferenceCount() == 1 and lConstValue == 0x0000FFFFFFFFFFF0

We do handle this case. We would catch it with this `OR` : `(lZeros > 0 && tZeros > 0)`. The reference count is irrelevant here. We will generate RISBG for this case no matter what. 

So if our lConstValue satisfies the pattern that the outer if-clause is trying to detect AND if the refCount>1 then it's ALWAYS better to generate a RISBG instruction. The alternative to a RISBG is multiple instructions (back up the value, then use destructive AND instruction).

>The best of both worlds is to determine in which case can we NOT use one of the NI** variants to evaluate the sequence when firstChild->getReferenceCount() == 1 and generate an RISBG in such cases (if possible of course).

We do essentially do that. The inner-if will only allow a RISBG to be generated if the alternative is to generate multiple `ANDs`. The refCount>1 check is done for the reasons explained above. The remaining checks in this clause can be explained by the following explanation:
>I think a good way to think about this is that we are better off generating a RISBG if there's a 0 in the top half AND the bottom half of the constant (given that we have the leadingOnes or leadingZeros pattern). A 0 means that we need to explicitly zero out a particular bit in the register value if we use these AND Immediate instructions. If the zero exists in both halves then we will need two AND Immediate instructions.

fjeremic replied:
Thanks for the explanation. My mistake was that I was reading the code as:

```
if (firstChild->getReferenceCount())
   {
   ...
   }
```

fjeremic commented on compiler/z/codegen/BinaryEvaluator.cpp:
```diff
@@ -2476,56 +2504,75 @@ genericLongAndAsRotateHelper(TR::Node * node, TR::CodeGenerator * cg)
-                  || lZeros > 31 || tZeros > 31 || (lZeros > 0 && tZeros > 0))  // Better off with a single AND?
+            if (firstChild->getReferenceCount() > 1 // if we use an AND and refCount>1, then we will have to backup the value in another register beforehand, as AND is destructive
+                || lZeros > 31 || tZeros > 31 // this makes sure we are not better off using one of the following AND instructions: NILF,NILH,NILL, NIHF,NIHH, NIHL
+                || (lZeros > 0 && tZeros > 0)) // if our constant at minimum looks like 0111...110, then it's better to use RISBG as this case will require multiple ANDs
```
This does not look correct. What if `lZeros == 1` and `tZeros == 60`. In this case you would carry out the transformation even though you could have generated `NILL` which is a 4-byte instruction as opposed to the 6-byte `RISBG`. This needs more careful consideration.

z/Architecture has all variants of AND immediate instructions for the 16-bit non-overlapping parts of a register. We should only be using `RISBG` in the cases where we cannot handle the AND with a single such 16-bit AND immediate operation. We need to have checks for this.

dchopra001 replied:
So I don't thing that example would work. A value such as that looks like this in binary:
011100000...0. We would need to explicitly zero out everything except those three bits. As per the specification for the 6 AND immediate instructions in Z, 

>"the second operand is ANDed with bits of the first operand, and the result replaces those bits of the first operand. The remainder of the operand remains unchanged."

The 6 AND immediate operands are below:

![andimm](https://user-images.githubusercontent.com/11260919/32392454-333b3c36-c0ac-11e7-8fbb-1a5221b039cc.PNG)

The problem is in the fact that the remainder of the operand remains unchanged. So we can't use NILL because that would mean bits 0-47 would remain unchanged when they should be explicitly zeroed out. In fact we would probably need to generate a NIHF and NILF instruction because we need to make changes to the entire top half and bottom half of the register. 

I think a good way to think about this is that we are better off generating a RISBG if there's a 0 in the top half AND the bottom half of the constant (given that we have the leadingOnes or leadingZeros pattern). A 0 means that we need to explicitly zero out a particular bit in the register value if we use these AND Immediate instructions. If the zero exists in both halves then we will need two AND Immediate instructions. 

What you're suggesting does work for the converse case though. If we had something like 1000111111...111 instead, then we could use NILL to zero out those three bits and keep the rest of the bits in the register unchanged. This is exactly why the `if` clause on line 2569 seems a lot less restrictive than the one here because it can handle that case.

Perhaps this also deserves a better comment?

fjeremic commented on compiler/z/codegen/BinaryEvaluator.cpp:
```diff
@@ -2476,56 +2504,75 @@ genericLongAndAsRotateHelper(TR::Node * node, TR::CodeGenerator * cg)
                }
-            }
-         else if (popCnt == (lOnes + tOnes))
+            }                         
```
Unnecessary space here.

fjeremic commented on compiler/z/codegen/BinaryEvaluator.cpp:
```diff
@@ -2476,56 +2504,75 @@ genericLongAndAsRotateHelper(TR::Node * node, TR::CodeGenerator * cg)
-            }
-         else if (popCnt == (lOnes + tOnes))
+            }                         
+         else if (popCnt == (lOnes + tOnes) && popCnt > 0) 
```
In terms of compile time would it be better to first compute `popCnt` and have an early exit if `popCnt == 0`?

dchopra001 replied:
Yeah it's better to pull this check up and bail as soon as we calculate that popCnt is 0. It will make more sense as well when someone is reading this code in the future. I'll change this in the next commit.

fjeremic commented on compiler/z/codegen/BinaryEvaluator.cpp:
```diff
@@ -2476,56 +2504,75 @@ genericLongAndAsRotateHelper(TR::Node * node, TR::CodeGenerator * cg)
             lsBit = lOnes - 1;
-            if (firstChild->getReferenceCount()>1 || skipedConversion != NULL // Better of with non-destructive RISBG as AND?
-                  || (lOnes < 32 && tOnes < 32) )                             // Better off with a single AND?
+            if (firstChild->getReferenceCount()>1 // if we use an AND and refCount>1, then we will have to backup the value in another register beforehand, as AND is destructive
```
Same as above. Can we not use clobber evaluate if it will save generating two ANDs?

dchopra001 replied:
As mentioned in another comment, I don't think we need clobber evaluate. We don't need to store the original value in another register because the RISBG instruction is not destructive. So we can just store the result of the RISBG instruction in another register. I would think a regular evaluate is probably faster in terms of compile time performance as well?

fjeremic replied:
> I would think a regular evaluate is probably faster in terms of compile time performance as well?

Yes we just need a regular evaluation of the first child. I don't think we need to reference count check here as a guard.

fjeremic commented on compiler/z/codegen/BinaryEvaluator.cpp:
```diff
@@ -2539,7 +2586,12 @@ genericLongAndAsRotateHelper(TR::Node * node, TR::CodeGenerator * cg)
+               // The bits in that range are preserved as is
                generateRIEInstruction(cg, opCode, node, targetReg, sourceReg, msBit, 0x80 + lsBit, 0);
 
             if (skipedConversion)
```
This looks like dead code and should be removed.

fjeremic commented on compiler/z/codegen/BinaryEvaluator.cpp:
```diff
@@ -4725,7 +4776,7 @@ OMR::Z::TreeEvaluator::landEvaluator(TR::Node * node, TR::CodeGenerator * cg)
    if ((TR::Compiler->target.is64Bit() || cg->use64BitRegsOn32Bit()) &&
        ((targetRegister = genericRotateAndInsertHelper(node, cg))
-             || (targetRegister = genericLongAndAsRotateHelper(node, cg)))     )
+             || (targetRegister = tryToReplaceLongAndWithRotateInstruction(node, cg)))     )
```
Can we fix the spacing here on the last parenthesis?

fjeremic commented on fvtest/compilertriltest/CMakeLists.txt:
```diff
@@ -39,6 +39,7 @@ add_executable(comptest
 	ShiftAndRotateTest.cpp
 	SimplifierFoldAndTest.cpp
 	IfxcmpgeReductionTest.cpp
+        LongAndAsRotateTest.cpp
```
I think the CMakeLists.txt files use tabs, hence the spacing inconsistency here. Can we modify the spaces to tabs?

rbanerjee commented on compiler/z/codegen/BinaryEvaluator.cpp:
```diff
@@ -2476,56 +2504,75 @@ genericLongAndAsRotateHelper(TR::Node * node, TR::CodeGenerator * cg)
-            firstChild = firstChild->getFirstChild();
-            value &= 0x00000000FFFFFFFFL;
-            }
+            TR::DebugCounter::incStaticDebugCounter(cg->comp(), TR::DebugCounter::debugCounterName(cg->comp(), "codegen/z/BinaryEvaluator/simplifierShouldHaveFoldedThisTree/%s", cg->comp()->signature()));
```
Kind of a small nit, but do you mind wrapping the lines for the newly added code. I believe we don't have a limit on line length, but some of the line lengths here do seem inconsistent with the convention followed around them.

rbanerjee commented on compiler/z/codegen/BinaryEvaluator.cpp:
```diff
@@ -2476,56 +2504,75 @@ genericLongAndAsRotateHelper(TR::Node * node, TR::CodeGenerator * cg)
             lsBit = lOnes - 1;
-            if (firstChild->getReferenceCount()>1 || skipedConversion != NULL // Better of with non-destructive RISBG as AND?
-                  || (lOnes < 32 && tOnes < 32) )                             // Better off with a single AND?
+            if (firstChild->getReferenceCount()>1 // if we use an AND and refCount>1, then we will have to backup the value in another register beforehand, as AND is destructive
```
Please see the comment above about wrapping really long lines.

rbanerjee commented on fvtest/compilertriltest/LongAndAsRotateTest.cpp:
```diff
@@ -0,0 +1,494 @@
+         "  (land                            " 
+         "      (i2l (iload parm=0))         "
+         "      (lconst %s)))))              ",
+         std::to_string(param.constantValue).c_str()
```
Sorry but this isn't going to work on our build compiler. 

I believe `std::to_string` only got a constructor overload for `int` starting `C++11`. 

A possible workaround would be to do the following :

```
...

(lconst %lld) ))))) 

...
param.constantValue

```

Please do the same for the rest of the cases where you are using ` std::to_string(param.constantValue).c_str()` 
is used.

rbanerjee commented on fvtest/compilertriltest/LongAndAsRotateTest.cpp:
```diff
@@ -0,0 +1,494 @@
+         "(block(lstore temp=\"x\"(land(i2l id=\"myNode\"      "
+         "(iload parm=0))(lconst 0xFFFFFFFFFFFF0000)))         "
+         "(lreturn(land(@id \"myNode\")(lconst %s)))))         ",  
+         std::to_string(param.constantValue).c_str()
```
See comment about the `std::to_string` constructor.

rbanerjee commented on fvtest/compilertriltest/LongAndAsRotateTest.cpp:
```diff
@@ -0,0 +1,494 @@
+         "(lconst 0xFFFFFFFFFFFF0000)))                          "
+         "(lreturn(land(i2l(iand(@id \"myNode\")(iconst 0x1234)))"
+         "(lconst %s)))))                                        ",  
+         std::to_string(param.constantValue).c_str()
```
Same as above.

rbanerjee commented on fvtest/compilertriltest/LongAndAsRotateTest.cpp:
```diff
@@ -0,0 +1,494 @@
+         "         )                         "
+         "      )                            " 
+         "      (lconst %s)))))              ",
+         std::to_string(param.constantValue).c_str()
```
Same as above.

rbanerjee commented on fvtest/compilertriltest/LongAndAsRotateTest.cpp:
```diff
@@ -0,0 +1,494 @@
+         "  (land                            " 
+         "      (iu2l (iload parm=0))         "
+         "      (lconst %s)))))              ",
+         std::to_string(param.constantValue).c_str()
```
Same as above.

rbanerjee commented on fvtest/compilertriltest/LongAndAsRotateTest.cpp:
```diff
@@ -0,0 +1,494 @@
+         "      )                                              "
+         "   )                                                 "
+         ")                                                    ",  
+         std::to_string(param.constantValue).c_str()
```
Same as above.

rbanerjee commented on fvtest/compilertriltest/LongAndAsRotateTest.cpp:
```diff
@@ -0,0 +1,494 @@
+         "(block(lstore temp=\"x\"(land(iu2l id=\"myNode\"      "
+         "(iload parm=0))(lconst 0xFFFFFFFFFFFF0000)))         "
+         "(lreturn(land(@id \"myNode\")(lconst %s)))))         ",  
+         std::to_string(param.constantValue).c_str()
```
Same as above.

rbanerjee commented on fvtest/compilertriltest/LongAndAsRotateTest.cpp:
```diff
@@ -0,0 +1,494 @@
+         "(lconst 0xFFFFFFFFFFFF0000)))                          "
+         "(lreturn(land(iu2l(iand(@id \"myNode\")(iconst 0x1234)))"
+         "(lconst %s)))))                                        ",  
+         std::to_string(param.constantValue).c_str()
```
Same as above.

rbanerjee commented on fvtest/compilertriltest/LongAndAsRotateTest.cpp:
```diff
@@ -0,0 +1,494 @@
+         "         )                         "
+         "      )                            " 
+         "      (lconst %s)))))              ",
+         std::to_string(param.constantValue).c_str()
```
Same as above.

dchopra001 commented on fvtest/compilertriltest/LongAndAsRotateTest.cpp:
```diff
@@ -0,0 +1,494 @@
+#include <Jit.hpp>
+
+// these two vectors hold the value loaded by the iload node.
+std::vector<int32_t> i2lInput = {0x0,0x1,0x0000FFFF,0x2,0x3E,(int32_t)0x80000000,(int32_t)0xFFFF0000,0x40000000,0x7C000000,(int32_t)0xFFFFFFFF,0x0003C000,0x40000001};
```
How about I just use a c style array then to avoid all confusion?

```int32_t i2lInput [] = {0x0,0x1,0x0000FFFF,0x2,0x3E,(int32_t)0x80000000,
                         (int32_t)0xFFFF0000,0x40000000,0x7C000000,
                         (int32_t)0xFFFFFFFF,0x0003C000,0x40000001};```

Leonardo2718 replied:
Yep, that might work.

dchopra001 commented:
@fjeremic @mgaudet I've addressed all the comments/changes you requested. PTAL.

@rbanerjee @Leonardo2718 Thank you for the review. I'd missed the function pointer casting and vector initializations when downgrading the tests to be compatible with our compilers. I believe all the other changes you guys requested should be in this commit. PTAL and let me know if I missed something.

Thanks!

mgaudet commented:
High level inspection of the tests: LGTM.

0xdaryl commented:
genie-omr build all

0xdaryl commented:
@rbanerjee : are you OK to merge?  Please approve.

rbanerjee commented:
Looks Great! (Thanks for addressing the formatting issues as well) 😁
